crates/hc_sandbox/src/save.rs

```diff
@@ -4,85 +4,168 @@
+        )
+    })?;
+    let hc_file = hc_dir.join(".hc");
     let mut file = std::fs::OpenOptions::new()
         .append(true)
         .create(true)
-        .open(hc_dir)?;
-
+        .open(hc_file)
+        .map_err(|err| {
+            std::io::Error::new(
+                err.kind(),
+                format!(
+                    "Failed to create `.hc` file at '{}': {}",
+                    hc_dir.display(),
+                    err
+                ),
+            )
+        })?;
     for path in paths {
         writeln!(file, "{}", path.display())?;
     }
     Ok(())
 }
 
-/// Remove sandboxes by their index in the file.
-/// You can get the index by calling [`load`].
-/// If no sandboxes are passed in then all are deleted.
-/// If all sandboxes are deleted the `.hc` file will be removed.
-pub fn clean(mut hc_dir: PathBuf, sandboxes: Vec<usize>) -> anyhow::Result<()> {
-    let existing = load(hc_dir.clone())?;
-    let sandboxes_len = sandboxes.len();
-    let to_remove: Vec<_> = if sandboxes.is_empty() {
-        existing.iter().collect()
+/// Remove sandbox paths from the `.hc` file and attempt to delete the sandbox folders.
+/// If no sandbox paths remain in the `.hc` file, then the `.hc`, `.hc_auth`, and all `.hc_live*`
+/// files will be removed from `hc_dir`.
+/// Returns the number of removed paths from `.hc`.
+pub fn remove(hc_dir: PathBuf, existing: Existing) -> std::io::Result<usize> {
+    let sandboxes = load(hc_dir.clone())?;
+    // Determine sandbox paths to delete
+    let mut to_remove_indices: Vec<usize> = Vec::new();
+    if existing.all {
+        to_remove_indices = (0..sandboxes.len()).collect();
     } else {
-        sandboxes
+        existing
+            .indices
             .into_iter()
-            .filter_map(|i| existing.get(i))
-            .collect()
-    };
-    let to_remove_len = to_remove.len();
-    for p in to_remove.into_iter().flatten() {
+            .for_each(|i| match sandboxes.get(i) {
+                None => msg!("Warning: Provided index is out of range: {}", i),
+                Some(Err(path)) => msg!("Warning: Missing sandbox {}:{}", i, path.display()),
+                Some(Ok(_)) => to_remove_indices.push(i),
+            });
+    }
+    // Determine remaining paths
+    let indices_to_remove: HashSet<_> = to_remove_indices.iter().collect();
+    let remaining: Vec<ConfigRootPath> = sandboxes
+        .iter()
+        .enumerate()
+        .filter_map(|(i, item)| {
+            if indices_to_remove.contains(&i) {
+                return None;
+            }
+            let Ok(item) = item else {
+                return None;
+            };
+            Some(ConfigRootPath::from(item.clone()))
+        })
+        .collect();
+    // Attempt to delete each sandbox
+    for i in to_remove_indices.iter() {
+        let Some(Ok(p)) = sandboxes.get(*i) else {
+            continue;
+        };
         if p.exists() && p.is_dir() {
```

Comment:
Related to my comment above, we check here anyway whether the directory exists.

Reply:
Its not the case when calling remove with `Existing::all`. Dirs are not checked.

Reply:
I removed the check since the user will get a clean error message anyway.